File: sample/src/main/java/ir/farshid_roohi/customadapterrecyclerview/MainActivity.java
package ir.farshid_roohi.customadapterrecyclerview;

import android.os.Bundle;
import android.os.Handler;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import ir.farshid_roohi.customadapterrecycleview.listener.OnClickItemListener;
import ir.farshid_roohi.customadapterrecycleview.listener.OnLoadMoreListener;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        RecyclerView recyclerView = findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new GridLayoutManager(this, 2));

        final MyAdapter adapter = new MyAdapter();
        adapter.endLessScrolled(recyclerView);
        adapter.addItems(getTempItems());
        recyclerView.setAdapter(adapter);


        // handled click item in recyclerView
        adapter.setOnClickItemListener(recyclerView, new OnClickItemListener<String>() {
            @Override
            public void onClickItem(int position, String element) {
                Toast.makeText(getApplicationContext(), "item click : " + element, Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onLongClickItem(int position, String element) {
                Toast.makeText(getApplicationContext(), "item long click : " + element, Toast.LENGTH_SHORT).show();

            }
        });


        // handled endless recyclerView
        adapter.setOnLoadMoreListener(new OnLoadMoreListener() {
            @Override
            public void onLoadMore() {
                adapter.showLoading();

                // fake request or load other items...
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        adapter.addItems(getTempItems());
                    }
                }, 2000);
            }
        });
    }

    // get temp list item
    public List<String> getTempItems() {
        List<String> items = new ArrayList<>();

        for (int i = 0; i <= 15; i++) {
            items.add("item ");
        }

        return items;
    }
}

File: sample/src/main/java/ir/farshid_roohi/customadapterrecyclerview/MainActivity.kt
package ir.farshid_roohi.customadapterrecyclerview

import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import ir.farshid_roohi.customadapterrecycleview.extensions.onItemClickListener
import ir.farshid_roohi.customadapterrecycleview.extensions.onLoadMoreListener
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val myAdapter = MyAdapter()

        // onClick Error Button
        myAdapter.onRetryClicked = {
            myAdapter.loadingState()
            // fake request or load other items...
            recyclerView.postDelayed({
                myAdapter.loadedState(tempItems)
            }, 2000)
        }

        // set Data in List
        myAdapter.loadedState(tempItems)

        recyclerView.layoutManager = GridLayoutManager(this@MainActivity, 2)
        recyclerView.adapter = myAdapter


        // set onLoad More Listener for Pagination
        recyclerView.onLoadMoreListener {
            if (myAdapter.mustLoad) {
                myAdapter.loadingState()
                // failed state
                recyclerView.postDelayed({
                    myAdapter.failedState()
                }, 2000)
            }
        }


        // Set onClick Item Listener
        recyclerView.onItemClickListener({ position ->
            if (position == myAdapter.itemCount - 1 && !myAdapter.mustLoad) {
                return@onItemClickListener
            }
            Toast.makeText(applicationContext, "item click :  ${myAdapter.getItem(position)}$position", Toast.LENGTH_SHORT).show()
        }, { position ->
            if (position == myAdapter.itemCount - 1 && !myAdapter.mustLoad) {
                return@onItemClickListener
            }
            Toast.makeText(applicationContext, "item long click :  ${myAdapter.getItem(position)}$position", Toast.LENGTH_SHORT).show()
        })
    }


    // Generate Fake Data For List
    private val tempItems: List<String>
        get() {
            val items: MutableList<String> = ArrayList()
            for (i in 0..21) {
                items.add("item ")
            }
            return items
        }
}